### PR TITLE
Add GLM-ASR Nano speech-language support

### PR DESCRIPTION
**Pull request by justinchuby**
## Summary

- Add complete `glmasr` support for `zai-org/GLM-ASR-Nano-2512` as standardized `audio_encoder` → `embedding` → cached `decoder` ONNX models.
- Implement the convolutional 32-layer bidirectional partial-RoPE audio tower, config-driven audio/projector activations, four-frame projector, audio-token embedding replacement, and 28-layer Llama-like decoder with dynamic KV cache.
- Add exact checkpoint weight alignment, registry/config/task exports, ONNX GenAI metadata, immutable CLI `--revision` propagation, and documentation metadata.
- Add a pinned real nonzero speech fixture plus exact L4 prefill and L5 transcript/token goldens.

## Immutable inputs

- Model: `zai-org/GLM-ASR-Nano-2512@61ba4e0b3309b6656edea3e93e419f7bd5c61957`
- Audio: `eustlb/audio-samples@1045238c3afdc0254472ecdde5dae95370d621e4`
- All config, weight, tokenizer, processor, metadata, golden-generation, and test Hub calls carry the immutable revision.

## Architecture checklist

- [x] Nested GLM-ASR text/audio config extraction
- [x] Registry and default speech-language task wiring
- [x] Standardized `audio_encoder`, `embedding`, and `decoder` package keys
- [x] Whisper feature-extractor contract: float32 `[B,128,3000]`, 16 kHz, FFT 400, hop 160
- [x] 20-head bidirectional audio Attention with 32/64 partial RoPE, asymmetric Q/V/O biases, and one output (no unused present KV)
- [x] Config-driven audio MLP and projector activations
- [x] Four-frame `5120 → 4096 → 2048` projector
- [x] Exact audio-placeholder replacement and valid projected-feature stripping inside ONNX
- [x] Llama decoder with 16 query heads, 4 KV heads, full RoPE, and cached decode
- [x] All 745 checkpoint parameter names aligned after preprocessing
- [x] ONNX GenAI speech-language metadata includes the pinned audio processor

### L1–L3 and regression gates

- GLM-ASR graph/config/parity and rebase regression suite: **46 passed**
- ORT GenAI compatibility guard tests: **2 passed**
- Complete CLI suite: **37 passed**
- Synthetic FP32 HF/ONNX parity executes the complete audio → embedding → decoder pipeline and compares every stage, including short-audio zero-token behavior.
- Standard non-integration gate: **4051 passed, 58 skipped, 10 deselected**. Five unrelated environment-sensitive failures remain in GGUF Q4 dequantization, Qwen Image CUDA tolerance, and softcap Attention-decomposition parity; none overlap this change set or the required CI failures previously reported.

### L4 CUDA prefill (rebased head)

```powershell
$env:MOBIUS_TEST_DEVICE='cuda'
python -m pytest tests/e2e_golden_test.py -m golden --models glmasr -v --timeout=300 --tb=short
```

- **1 passed, 240 skipped, 161 deselected in 166.73s**
- Real pinned speech, 138 projected audio tokens, 153-token prompt.
- FP16 full-logit reference comparison: shape `[1,153,59264]`, correlation `0.9991001`, mean absolute difference `0.02795`, maximum absolute difference `7.456`; greedy token matches.

### L5 CUDA cached generation (rebased head)

```powershell
$env:MOBIUS_TEST_DEVICE='cuda'
python -m pytest tests/e2e_golden_test.py -m generation --models glmasr -v --timeout=300 --tb=short
```

- **1 passed, 156 skipped, 245 deselected in 238.19s**
- Exact 23-token sequence and exact decoded transcript asserted:
  - `Yesterday it was thirty five degrees in Barcelona, but today the temperature will go down to minus twenty degrees.`

### CLI export and CUDA execution

```powershell
python -m mobius build --model zai-org/GLM-ASR-Nano-2512 <output> \
  --revision 61ba4e0b3309b6656edea3e93e419f7bd5c61957 \
  --ep cuda --dtype f16
```

- Direct three-model ONNX export and saved-file CUDA execution pass:
  - audio `(138, 2048)`
  - embeddings `(1, 153, 2048)`
  - logits `(1, 153, 59264)`
  - prefill top-1 token `89`
- The staged test harness now destroys each ORT session before loading the next component. This fixes the old CI `BFCArena` initialization OOM (`242745344`-byte decoder allocation) and returns CUDA memory to baseline between stages.

### Formatting and review

- `lintrunner -a`: **passed**
- Final specialist re-review: **no high-confidence findings**
- All inline review threads are addressed and resolved.

## Waivers and runtime constraints

- **FP32:** complete synthetic stage-by-stage semantic parity passes at `rtol=2e-4`, `atol=2e-4` for decoder logits and tighter tolerances for earlier stages. A full real FP32 CUDA checkpoint run is not practical on the available 8 GB RTX A1000.
- **FP16:** full real-checkpoint CUDA L4/L5 and saved CLI artifact execution pass as documented above.
- **BF16:** the weighted package builds and loads on CUDA. ORT 1.26 Python's NumPy API cannot materialize BF16 outputs (`No corresponding Numpy type for Tensor Type. bfloat16`), so executable L4/L5 uses FP16; BF16 output validation requires I/O binding or a runtime with BF16 host materialization.
- **GQA fusion:** CUDA FP16 `GroupQueryAttention` changed greedy termination from the exact 23-token HF sequence to an earlier valid EOS at 21 tokens. GLM-ASR intentionally preserves standard opset-24 `Attention` for exact transcript parity; graph coverage locks this dispatch decision.
- **ORT GenAI:** onnxruntime-genai 0.15.2 registers multimodal model types only for `gemma4` and `phi4mm`; loading `type: glmasr` fails as unsupported. Mobius now rejects `--runtime ort-genai` before saving rather than emitting unloadable metadata. Direct ONNX and ONNX GenAI metadata remain supported; native ORT GenAI packaging is deferred until the runtime registers a compatible GLM-ASR pipeline.
- **Memcpy:** ORT reports small control-tensor Memcpy nodes from dynamic shape/index operations (`NonZero`, integer `CumSum`, and placeholder indexing). Eliminating them would require changing the feature-stripping/embedding contract.

### CODE REVIEW

File: src/mobius/models/glm_asr.py [hunk 1]
```diff
@@ -0,0 +1,455 @@
+# Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/glm_asr.py:1]: Applied the repository formatter and confirmed `lintrunner -a` is clean on the rebased head.

File: src/mobius/models/glm_asr_test.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
+# Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/glm_asr_test.py:1]: Applied the Ruff fixes and confirmed `lintrunner -a` is clean on the rebased head.

[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/glm_asr_test.py:1]: Applied the repository formatter and confirmed `lintrunner -a` is clean on the rebased head.

File: src/mobius/models/glm_asr_test.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+from mobius._builder import build_from_module
+from mobius._configs import GlmAsrConfig
+from mobius._testing.ort_inference import OnnxModelSession
+from mobius.models.glm_asr import GlmAsrForConditionalGeneration
+from mobius.tasks import GlmAsrSpeechLanguageTask
+
+
+def _tiny_hf_model():
+    from transformers import (
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/glm_asr_test.py:19]: Sorted the local Transformers imports with Ruff and confirmed `lintrunner -a` is clean.

File: src/mobius/models/glm_asr.py [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+            value_states,
+            None,
+            None,
+            None,
+            q_num_heads=self._num_attention_heads,
+            kv_num_heads=self._num_key_value_heads,
+            scale=self._scale,
+            is_causal=0,
+            _outputs=3,
+        )
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/glm_asr.py]: Changed the bidirectional encoder to request only the Attention output and added a graph assertion that every audio Attention node has exactly one output. This also avoids materializing unused present KV tensors.

File: src/mobius/models/glm_asr.py [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+class GlmAsrAudioMLP(nn.Module):
+    """Bias-enabled GELU feed-forward network used by the audio encoder."""
+
+    def __init__(self, config: ArchitectureConfig):
+        super().__init__()
+        self.fc1 = Linear(config.hidden_size, config.intermediate_size, bias=True)
+        self.fc2 = Linear(config.intermediate_size, config.hidden_size, bias=True)
+
+    def forward(self, op: OpBuilder, hidden_states: ir.Value) -> ir.Value:
+        return self.fc2(op, op.Gelu(self.fc1(op, hidden_states)))
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/glm_asr.py]: Fixed both sites: the audio MLP now resolves `config.hidden_act`, and the projector resolves `config.projector_hidden_act`, through the shared `get_activation()` helper. Added graph coverage using non-default activations.

**Comment by justinchuby** [comment]
@copilot please update from main
